### PR TITLE
POC implementation of deterministic GPU histogram.

### PR DESCRIPTION
**Pull request by trivialfis**
* Use pre-rounding based method to obtain reproducible floating point
  summation.
* This makes GPU Hist more stable but not yet fully reproducible.

### CODE REVIEW

**Comment by trivialfis** [comment]
Experimenting on Year Prediction Dataset:

## Non-deterministic histogram build:
non-deterministic

## Deterministic histogram build:
deterministic

File: tests/cpp/common/test_gpu_hist_util.cu [hunk 1]
```diff
@@ -76,6 +76,20 @@ void TestDeviceSketch(bool use_external_memory) {
     ASSERT_LT(fabs(hmat_cpu.Values()[i] - hmat_gpu.Values()[i]), eps * nrows);
   }
 
+  // Determinstic
```
Comment (trivialfis) [inline tests/cpp/common/test_gpu_hist_util.cu]: Note: Typo

File: src/tree/gpu_hist/histogram.cu [hunk 2]
```diff
@@ -0,0 +1,156 @@
...
+/* \brief Constructs a rounding factor used to truncate elements in a sum such that the
+   sum of the truncated elements is the same no matter what the order of the sum is.
+
+ * Follows Algorithm 5: Reproducible Sequential Sum in 'Fast Reproducible Floating-Point
+ * Summation' by Demmel and Nguyen
+
+ * @param max max(sum(positive gradient) + sum(abs(negative gradient)))
+ * @param n The number of elements for the sum, or an upper bound estimate
+ */
+DEV_INLINE __host__ float CreateRoundingFactor(float max, int n) {
```
Comment (RAMitchell) [inline src/tree/gpu_hist/histogram.cu]: I was expecting you to do this with doubles, this would give an extra 29 bits of precision and might resolve your accuracy issues. We can make the check to only do the truncation if GradientSumT is double precision.

**Comment by trivialfis** [comment]
Em I still get some non-determinism somewhere.  Need more investigation.

**Comment by trivialfis** [comment]
I was able to obtain bit by bit reproducible result on 1M rows HIGGS dataset.  But started to get discrepancy on ~5M YearPrediction.

**Comment by trivialfis** [comment]
There's minor discrepancy on 5M YearPrediction on may laptop but not on my desktop.   On 1M+ Higgs both have discrepancy, arguably much smaller than non-deterministic hist.

I added a new parameter so that we can switch if on certain datasets the accuracy drops too much.  But so far on YearPrediction and Higgs the accuracy loss is negligible.

**Comment by trivialfis** [comment]
@RAMitchell After getting `dh::SumReduction` out of the way, I was able to obtain bit by bit reproducible result on both YearPrediction, Higgs and Epsilon for 100 iterations.  As for why `cub::DeviceReduce` produces non-deterministic result I'm not sure.

**Comment by RAMitchell** [comment]
Awesome! Maybe uninitialised memory in `node_sum_gradients_d` or someone else?

**Comment by trivialfis** [comment]
@RAMitchell 

> Maybe uninitialised memory in `node_sum_gradients_d`

That's not the cause.

**Comment by trivialfis** [comment]
@RAMitchell  Time for training on Year Prediction is up from `1.46069` to `1.6410` for 100 iterations.  I don't have more commits coming unless there's further optimization required.  Please review when time allows.

**Comment by trivialfis** [comment]
Actually maybe I can pre-compute the rounding.

**Comment by mli** [comment]
# [Codecov](https://codecov.io/gh/dmlc/xgboost/pull/5361?src=pr&el=h1) Report
> Merging [#5361](https://codecov.io/gh/dmlc/xgboost/pull/5361?src=pr&el=desc) into [master](https://codecov.io/gh/dmlc/xgboost/commit/655cf17b60538a8655ca2abff29c74d3ce3a995c&el=desc) will **increase** coverage by `0.31%`.
> The diff coverage is `n/a`.

[Impacted file tree graph](https://codecov.io/gh/dmlc/xgboost/pull/5361?src=pr&el=tree)

```diff
@@            Coverage Diff             @@
##           master    #5361      +/-   ##
==========================================
+ Coverage   83.75%   84.07%   +0.31%     
==========================================
  Files          11       11              
  Lines        2413     2411       -2     
==========================================
+ Hits         2021     2027       +6     
+ Misses        392      384       -8     
```


| [Impacted Files](https://codecov.io/gh/dmlc/xgboost/pull/5361?src=pr&el=tree) | Coverage Δ | |
|---|---|---|
| [python-package/xgboost/compat.py](https://codecov.io/gh/dmlc/xgboost/pull/5361/diff?src=pr&el=tree#diff-cHl0aG9uLXBhY2thZ2UveGdib29zdC9jb21wYXQucHk=) | `54.54% <0.00%> (+0.58%)` | :arrow_up: |
| [python-package/xgboost/core.py](https://codecov.io/gh/dmlc/xgboost/pull/5361/diff?src=pr&el=tree#diff-cHl0aG9uLXBhY2thZ2UveGdib29zdC9jb3JlLnB5) | `80.10% <0.00%> (+0.65%)` | :arrow_up: |

------

[Continue to review full report at Codecov](https://codecov.io/gh/dmlc/xgboost/pull/5361?src=pr&el=continue).
> **Legend** - [Click here to learn more](https://docs.codecov.io/docs/codecov-delta)
> `Δ = absolute <relative> (impact)`, `ø = not affected`, `? = missing data`
> Powered by [Codecov](https://codecov.io/gh/dmlc/xgboost/pull/5361?src=pr&el=footer). Last update [655cf17...e8348f0](https://codecov.io/gh/dmlc/xgboost/pull/5361?src=pr&el=lastupdated). Read the [comment docs](https://docs.codecov.io/docs/pull-request-comments).

File: src/tree/gpu_hist/histogram.cu [hunk 3]
```diff
@@ -0,0 +1,155 @@
...
+GradientPair CreateRoundingFactor(common::Span<GradientPair> gpair) {
+  auto positive = []__device__(float v) {
+    return v > 0 ? v : 0;
+  };  // NOLINT
+  auto negative = []__device__(float v) {
+    return v < 0 ? abs(v) : 0;
+  };  // NOLINT
+
+  dh::XGBCachingDeviceAllocator<char> alloc;
+  auto positive_sum = thrust::reduce(
```
Comment (RAMitchell) [inline src/tree/gpu_hist/histogram.cu]: Not sure if it matters, but it is possible to fuse these two reductions into a single reduction e.g. apply transform iterator to gradient pair inputs that returns 2 GradientPair (pos/neg) and define an operator to sum them. Alternatively just use the absolute sum, although this is slightly less tight and we could lose up to one bit of precision.

File: src/tree/gpu_hist/histogram.cuh [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+#define HISTOGRAM_CUH_
+#include <thrust/transform.h>
+#include "../../data/ellpack_page.cuh"
+
+namespace xgboost {
+namespace tree {
+
+GradientPair CreateRoundingFactor(common::Span<GradientPair> gpair);
+
+DEV_INLINE float TruncateWithRoundingFactor(float rounding_factor, float x) {
```
Comment (RAMitchell) [inline src/tree/gpu_hist/histogram.cuh]: I don't understand how this function could be so expensive that it slows down the algorithm if we use it in the histogram kernel.

File: src/tree/gpu_hist/histogram.cu (hunk 3 above)
Comment (trivialfis) [reply src/tree/gpu_hist/histogram.cu]: @RAMitchell I tried using `cub::DeviceReudce` with 2 cuda streams for this, the end result doesn't seem to be any different.  As for abs sum, it failed a Windows test for GPU ranking when using float.

File: src/tree/gpu_hist/histogram.cuh (hunk 4 above)
Comment (trivialfis) [reply src/tree/gpu_hist/histogram.cuh]: @RAMitchell I pushed the change for computing truncation on the fly.  There are other functions like `GetHess`, `GetGrad` and constructors.  I'm not sure if the compiler can optimize them away.

**Comment by trivialfis** [comment]
> I think we need a Python level integration test.

Done.

**Comment by RAMitchell** [comment]
@sh1ng @pseudotensor you might find this feature useful.
